### ISSUE TITLE
Standalone Spark Executors on Kubernetes throw "Connection Timeout"

### ISSUE THREAD

**Issue by ursuad** [kubernetes-incubator/application-images#10]
Currently we're running Spark on K8s in AWS with Flannel. 

**Spark Version**: 1.6.2
**Spark Deployment Mode**: Standalone
**K8s Version**: 1.3.7

The cluster is up, and all the components connect successfully. The issues appear when we submit a job to Spark.  
In the Spark UI, when I go to the executors tab, I see wrong IP address for executors, which doesn't match the POD IP addresse. 

**POD IP Addresses** from kubectl

`$kubectl get po`

Name | IP
------------ | -------------
spark-master-7v08x | 10.70.20.5
spark-worker-1soa1 | 10.70.19.6
spark-worker-mg7q5 | 10.70.28.8
spark-worker-vvr4n | 10.70.20.4

**The executor IP Addresses** from the Spark UI

 Executor ID | Address
------------ | -------------
0 | 10.70.20.4:39955
1 | 10.70.20.4:48970
2 | 10.70.19.0:38067	

If I check the executor logs, I can see exceptions because the executors are trying to connect to the wrong IP:
```java
16/10/07 16:44:04 INFO RetryingBlockFetcher: Retrying fetch (2/3) for 1 outstanding blocks after 5000 ms
16/10/07 16:45:12 ERROR RetryingBlockFetcher: Exception while beginning fetch of 1 outstanding blocks (after 2 retries)
java.io.IOException: Failed to connect to /10.70.19.0:38067
	at org.apache.spark.network.client.TransportClientFactory.createClient(TransportClientFactory.java:216)
	at org.apache.spark.network.client.TransportClientFactory.createClient(TransportClientFactory.java:167)
	at org.apache.spark.network.netty.NettyBlockTransferService$$anon$1.createAndStart(NettyBlockTransferService.scala:90)
	at org.apache.spark.network.shuffle.RetryingBlockFetcher.fetchAllOutstanding(RetryingBlockFetcher.java:140)
	at org.apache.spark.network.shuffle.RetryingBlockFetcher.access$200(RetryingBlockFetcher.java:43)
	at org.apache.spark.network.shuffle.RetryingBlockFetcher$1.run(RetryingBlockFetcher.java:170)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.net.ConnectException: Connection timed out: /10.70.19.0:38067
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:224)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:289)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:528)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
	... 1 more
```

I've found some related issues on the Spark Jira:

* [Spark using Netty RPC gets wrong address in some setups](https://issues.apache.org/jira/browse/SPARK-14437)
* [shuffle broken when accessing standalone cluster through NAT](https://issues.apache.org/jira/browse/SPARK-14849)

I've seen that @mattf and @zmerlynn  did a lot of work around running spark in docker and on Kubernetes so I was hoping we can come up with some configuration workaround ( probably by using Kubernetes features) for launching the cluster which will prevent this from happening.

**Comment by erictune** [comment]
@foxish

**Comment by mattf** [comment]
for the web ui issue, you should wait for spark 2.1.0 and https://issues.apache.org/jira/browse/SPARK-15487

as for your connection issue, you should make sure your driver is running on kubernetes and thus the flannel managed network. if you MUST run the driver outside kubernetes, consider running it on kubernetes again and if you still need to there are options involving using NodePort on the workers or setting up a vpn into the flannel network.

**Comment by ursuad** [comment]
We run the driver in a container inside Kubernetes, so that's not a problem. 
Also the Spark UI is fully clickable so that's not a problem also. I launched [this ](https://github.com/aseigneurin/spark-ui-proxy) spark-ui-proxy as a quick workaround for that.

**Comment by mattf** [comment]
very nice find! with spark 2.1 you'll be able to replace that tunnel w/ a spark configuration change.

is this issue resolved? turn it into a request to update the application image for spark to v2.1?

**Comment by ursuad** [comment]
@mattf not really. The problem here is that the executor addresses have the wrong IPs which leads to  the specified `Connection timed out` in the executor's `stderr` affecting the deployed Spark Application.

I suppose a resolution would imply:
• understanding exactly why this is happening
• see if we can fix this by passing configuration to the spark workers when we launch the Pods

I've tried passing `SPARK_LOCAL_IP ` to the worker pod when it starts up, but I've seen [here](https://databricks.gitbooks.io/databricks-spark-knowledge-base/content/troubleshooting/connectivity_issues.html)  that:

> • If the `SPARK_LOCAL_IP` environment variable is set to a hostname, then this hostname will be used. If `SPARK_LOCAL_IP` is set to an IP address, it will be resolved to a hostname.
>• The IP address of the interface returned from Java's InetAddress.getLocalHost method.

[This](https://issues.apache.org/jira/browse/SPARK-14437) spark issue seems to be very similar with the behaviour that I'm seeing. Maybe it's better explained in there.

**Comment by foxish** [comment]
@ursuad There was an update to the example recently that you might want to try. https://github.com/kubernetes/kubernetes/pull/33604 which should be merged soon.

**Comment by ursuad** [comment]
Thanks @foxish, I tried those. Those changes unfortunately don't fix the issues that we're having.

**Comment by ursuad** [comment]
After doing some digging I found the issue. It's mentioned on the Spark Jira also [here](https://issues.apache.org/jira/browse/SPARK-14437).  
There is already a [merged PR](https://github.com/apache/spark/pull/12240) for Spark 2.0 (so that shouldn't be a problem) but it wasn't included in Spark 1.6.X versions. 

So for Spark 1.6 I needed to:
• checkout the Spark 1.6 code
• apply the PR
• build spark and pack it into a docker image. 

I'm closing this issue because there's nothing we can do from Kubernetes ( like customising config) to fix this. It's a Spark specific problem which has been fixed in Spark 2.0
